.emacs.d/lisp/package/packup.el

```diff
@@ -10,98 +10,149 @@
+a modified `package-archive-contents' that only includes the package-desc
+from the most-preferred repository."
+  (if (version< emacs-version "25.1")
+      `(let ((package-archive-contents
+              (mapcar (lambda (pkg)
+                        ;; first check package-pinned-packages
+                        (let* ((pinned-archive (cdr (assoc (car pkg) package-pinned-packages)))
+                               (desc (when pinned-archive
+                                       (car (cl-member pinned-archive (cdr pkg)
+                                                       :key #'package-desc-archive
+                                                       :test #'string=)))))
+                          (dolist (archive kotct/package-ordered-archives)
+                            (when (not desc)
+                              ;; no match yet, keep looking
+                              (setf desc (car (cl-member archive (cdr pkg)
+                                                         :key #'package-desc-archive
+                                                         :test #'string=)))))
+                          (list (car pkg) desc)))
+                      package-archive-contents)))
+         ,body)
+    `(progn ,body)))
+
+(defun kotct/package-latest-available (package)
+  "Get the lastest-available package-desc for PACKAGE, preferring repositories
+as listed in `package-archive-priorities' or `kotct/package-ordered-archives'.
+Does not automatically refresh the package list."
+  (if (version< emacs-version "25.1")
+      ;; no package-archive-priorities, so use kotct/package-ordered-archives
+      (let* ((versions (cdr (assoc package package-archive-contents)))
+             ;; check package-pinned-packages first
+             (pinned-archive (cdr (assoc package package-pinned-packages)))
+             (found-desc (when pinned-archive
+                     (car (cl-member pinned-archive versions
+                                     :key #'package-desc-archive
+                                     :test #'string=)))))
+        (dolist (archive kotct/package-ordered-archives)
+          (when (not found-desc)
+            ;; no match yet, keep looking
+            (setf found-desc (car (cl-member archive versions
+                                             :key #'package-desc-archive
+                                             :test #'string=)))))
+        found-desc)
+    ;; package-archive-priorities means we only have to check
+    ;; the first package-desc, because that should be from the best repo
+    (cadr (assoc package package-archive-contents))))
+
 (defun kotct/package-up-to-date-p (package)
-  "Returns true if PACKAGE is up-to-date.
-Does not automatically refresh package list."
-  (every (lambda (x) (package-installed-p package (package-desc-version x)))
-         (cdr (assq package package-archive-contents))))
+  "Returns T if PACKAGE is installed and up-to-date.
```

Comment:
`T` &rarr; `t`?

Reply:
Actually, it should probably just be "non-nil", since that's what `package-installed-p` says in its docstring.